File: Chapter11/com/packtpub/spark/ActivityOne.scala
package com.packtpub.spark.module_four.chapter_11

import org.apache.spark.sql.SparkSession

object ActivityOne {

  case class Saying(word: String)

  def main(args: Array[String]): Unit = {

    // Build a SparkSession in Local Mode
    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("My Spark App")
      .enableHiveSupport()
      .getOrCreate()

    // Import Spark Implicits for using .as[] function
    import spark.implicits._

    // Create some data from a funny saying with repetitious words in it
    val saying = "You cannot end a sentence with because, because because is a conjunction."

    // split the words by space so that they form a sequence of words
    val wordsOfTheSaying = saying.split(" ")

    // Consume the data into spark by parallelizing it and converting to a DataFrame (and optionally, a DataSet)
    val sayingInSpark = spark.sparkContext.parallelize(wordsOfTheSaying)
      .toDF("word")
      .as[Saying]

    sayingInSpark.show()

    // Clean the data by removing unneeded punctuation (by mapping over the data)
    val cleanedSaying = sayingInSpark.map(record => {

      // Store the word for continued use
      val wordOfSaying = record.word

      // if this word contains a comma, exclamation mark, or period, remove it!
      if (wordOfSaying.contains(",") || wordOfSaying.contains("!") || wordOfSaying.contains(".")){
        val cleanedWord = wordOfSaying
                          .replace(",", "")
                          .replace("!", "")
                          .replace(".", "")

        // Return the cleaned up word
        Saying(cleanedWord)
      }
      else {

        // No punctuation found, return the word as-is
        Saying(wordOfSaying)
      }
    })

    cleanedSaying.show()

    // Analyze the cleaned up Saying by counting the occurrence of each word within
    val analysis = cleanedSaying
      .groupBy("word")
      .agg("word" -> "count")
      .orderBy($"count(word)".desc)

    // Let's see those results!
    analysis.show()
  }
}

File: Chapter11/com/packtpub/spark/ActivityOne.py
from pyspark.sql import SparkSession
from pyspark.sql import Row
import string


# Clean the word of punctuation
def clean(word):
    return word.translate(translator)


# Create a Spark Session
spark = SparkSession \
    .builder \
    .appName("My Spark App") \
    .master("local[2]") \
    .getOrCreate()

# create spark contexts
sc = spark.sparkContext

# create a set of data
saying = "You cannot end a sentence with because, because because is a conjunction."
words = saying.split(" ")

# convert to RDD
wordsRDD = sc.parallelize(words)

# create a translator for removing punctuation and execute
translator = str.maketrans('', '', string.punctuation)
cleaned = wordsRDD.map(clean)

# convert rows to Row() format
rowRDD = cleaned.map(lambda x: Row(x))

# create dataframe
saying_in_spark = spark.createDataFrame(rowRDD, ['word'])

# analyze
analysis = saying_in_spark\
    .groupBy('word')\
    .agg({"word": "count"})\
    .sort("count(word)", ascending=False)

analysis.show()
